# Problem Statement
- This Dataset Contain the information about the bank customer that will take loan.
- Our goal is build a model that will detect weather the customer is eligible for loan or not.
- Data should be present in 2 different files.
- Both conatin same data about bank.
- Both files also contain `-99999` values.
- `-99999` are the missing values.

# Import Packages

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import warnings
warnings.filterwarnings('ignore')

# Read Data
1. Data is place another folder now find the path and read data.

In [2]:
file_path=[]
for i in os.listdir("Data"):
    file_path.append(f"{os.path.join(os.getcwd())}\Data\{i}")
    print(f"{os.path.join(os.getcwd())}\Data\{i}")

C:\Users\Laptop House\Weekly Task\CreditRisk\Data\case_study1.xlsx
C:\Users\Laptop House\Weekly Task\CreditRisk\Data\case_study2.xlsx


In [3]:
df1=pd.read_excel(file_path[0])
df2=pd.read_excel(file_path[1])

# Data Preprocessing
1. Top records
2. Shape of data
3. Check columns
4. Merge based on `common column`
5. Check Datatyes
6. Null values
7. Duplicates
8. Unique Values
9. Statistical Summary

# 1.Top Records

In [4]:
df1.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
35737       35738         1              0              1   
795           796         1              1              0   
12421       12422         5              1              4   
35899       35900         4              1              3   
50514       50515         2              0              2   
3414         3415         2              2              0   
817           818         2              0              2   
2924         2925         3              0              3   
9947         9948         1              1              0   
27670       27671         1              1              0   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
35737                    0                  0            0.000   
795                      0                  0            0.000   
12421                    3                  0            0.600   
35899                    1                  0            0.250   
50514                    1                  0            0.500   
3414                     0                  0            0.000   
817                      0                  0            0.000   
2924                     1                  0            0.333   
9947                     0                  0            0.000   
27670                    0                  1            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  CC_TL  \
35737                0.0           1.00           0.00  ...      0   
795                  0.0           0.00           1.00  ...      0   
12421                0.0           0.80           0.20  ...      0   
35899                0.0           0.75           0.25  ...      0   
50514                0.0           1.00           0.00  ...      0   
3414                 0.0           0.00           1.00  ...      0   
817                  0.0           1.00           0.00  ...      0   
2924                 0.0           1.00           0.00  ...      0   
9947                 0.0           0.00           1.00  ...      0   
27670                1.0           0.00           1.00  ...      0   

       Consumer_TL  Gold_TL  Home_TL  PL_TL  Secured_TL  Unsecured_TL  \
35737            0        0        0      0           1             0   
795              1        0        0      0           0             1   
12421            1        0        0      1           2             3   
35899            3        0        0      0           1             3   
50514            1        0        0      0           1             1   
3414             2        0        0      0           0             2   
817              1        0        0      0           1             1   
2924             1        0        0      2           0             3   
9947             1        0        0      0           0             1   
27670            1        0        0      0           0             1   

       Other_TL  Age_Oldest_TL  Age_Newest_TL  
35737         0             21             21  
795           0             38             38  
12421         1             28              3  
35899         1             32              3  
50514         0             23              6  
3414          0             22             18  
817           0             12              9  
2924          0             14              2  
9947          0             19             19  
27670         0              9              9  

[10 rows x 26 columns]

In [5]:
df2.sample(10)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
17055       17056                         75                       -99999   
8564         8565                        138                           14   
20330       20331                       1231                       -99999   
19772       19773                         62                            9   
8824         8825                        105                       -99999   
10038       10039                     -99999                       -99999   
46813       46814                     -99999                       -99999   
4249         4250                        143                           32   
19885       19886                     -99999                       -99999   
32955       32956                         75                       -99999   

       time_since_recent_deliquency  num_times_delinquent  \
17055                        -99999                     0   
8564                             12                     2   
20330                        -99999                     0   
19772                             2                     5   
8824                         -99999                     0   
10038                        -99999                     0   
46813                        -99999                     0   
4249                             32                     2   
19885                        -99999                     0   
32955                        -99999                     0   

       max_delinquency_level  max_recent_level_of_deliq  num_deliq_6mts  \
17055                 -99999                          0               0   
8564                      26                         26               0   
20330                 -99999                          0               0   
19772                     97                         68               1   
8824                  -99999                          0               0   
10038                 -99999                          0               0   
46813                 -99999                          0               0   
4249                      30                         30               0   
19885                 -99999                          0               0   
32955                 -99999                          0               0   

       num_deliq_12mts  num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  \
17055                0                  0  ...                     0.0   
8564                 0                  0  ...                     0.0   
20330                0                  0  ...                     0.0   
19772                5                  4  ...                     0.0   
8824                 0                  0  ...                     0.0   
10038                0                  0  ...                     0.0   
46813                0                  0  ...                     0.0   
4249                 0                  0  ...                     0.0   
19885                0                  0  ...                     0.0   
32955                0                  0  ...                     1.0   

       pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  \
17055                     1.0                     0.0   
8564                      0.0                     0.0   
20330                     0.0                     0.0   
19772                     0.0                     0.0   
8824                      0.0                     0.0   
10038                     0.0                     0.0   
46813                     0.0                     0.0   
4249                      0.0                     0.0   
19885                     0.0                     0.0   
32955                     1.0                     1.0   

       max_unsec_exposure_inPct  HL_Flag  GL_Flag  last_prod_enq2  \
17055                     2.196        0        0              PL   
8564                      2.000        0        0    ConsumerLoan   
20330      

# Conclussion
1. we can see that in `df2` most of the columns contain `-99999` values.
2. These `-99999` are basically `Missing value`.

# 2.shape

In [6]:
print("Data 1 Shape: ",df1.shape)
print("Data 2 Shape: ",df2.shape)

Data 1 Shape:  (51336, 26)
Data 2 Shape:  (51336, 62)


# Conclussion 
1. Data1 contain `51336` rows and `26` Column
2. Data2 contain `51336` rows and `62` Column

# 3.Columns

In [7]:
print(df1.columns)
print(df2.columns)

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL'],
      dtype='object')
Index(['PROSPECTID', 'time_since_recent_payment',
       'time_since_first_deliquency', 'time_since_recent_deliquency',
       'num_times_delinquent', 'max_delinquency_level',
       'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts',
       'num_deliq_6_12mts', 'max_deliq_6mts', 'max_deliq_12mts',
       'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std', 'num_std_6mts',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_6mts', 'num_dbt_

# Conclussion
1. See that both `dataframe` one common column `PROSPECTID`.
2. We can `merge` them first and then perform other data preprocessing.

# 4.Merge `df1` anf `df2`

In [8]:
df1.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  ...  CC_TL  Consumer_TL  Gold_TL  Home_TL  PL_TL  \
0            0.8  ...      0            0        1        0      4   
1            0.0  ...      0            1        0        0      0   

   Secured_TL  Unsecured_TL  Other_TL  Age_Oldest_TL  Age_Newest_TL  
0           1             4         0             72             18  
1           0             1         0              7              7  

[2 rows x 26 columns]

In [9]:
df2.head(2)

PROSPECTID  time_since_recent_payment  time_since_first_deliquency  \
0           1                        549                           35   
1           2                         47                       -99999   

   time_since_recent_deliquency  num_times_delinquent  max_delinquency_level  \
0                            15                    11                     29   
1                        -99999                     0                 -99999   

   max_recent_level_of_deliq  num_deliq_6mts  num_deliq_12mts  \
0                         29               0                0   
1                          0               0                0   

   num_deliq_6_12mts  ...  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0                  0  ...                     0.0                     0.0   
1                  0  ...                     0.0                     0.0   

   pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  HL_Flag  GL_Flag  \
0                     0.0                    13.333        1        0   
1                     0.0                     0.860        0        0   

   last_prod_enq2  first_prod_enq2  Credit_Score  Approved_Flag  
0              PL               PL           696             P2  
1    ConsumerLoan     ConsumerLoan           685             P2  

[2 rows x 62 columns]

In [10]:
final_df=df1.merge(df2,how='inner',on='PROSPECTID')
final_df.head(2)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  Total_TL_opened_L6M  \
0           1         5              4              1                    0   
1           2         1              0              1                    0   

   Tot_TL_closed_L6M  pct_tl_open_L6M  pct_tl_closed_L6M  pct_active_tl  \
0                  0              0.0                0.0            0.2   
1                  0              0.0                0.0            1.0   

   pct_closed_tl  ...  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
0            0.8  ...                     0.0                     0.0   
1            0.0  ...                     0.0                     0.0   

   pct_CC_enq_L6m_of_ever  max_unsec_exposure_inPct  HL_Flag  GL_Flag  \
0                     0.0                    13.333        1        0   
1                     0.0                     0.860        0        0   

   last_prod_enq2  first_prod_enq2  Credit_Score  Approved_Flag  
0              PL               PL           696             P2  
1    ConsumerLoan     ConsumerLoan           685             P2  

[2 rows x 87 columns]

# Conclussion
1. we have a `final df` now we can further work on this `final_df`

# 5. Datatypes

In [11]:
data_types=final_df.dtypes.to_dict()

In [12]:
pd.reset_option('display.max_rows')
pd.reset_option('display.max_columns')
pd.DataFrame(list(data_types.items()),columns=['Columns','Datatypes']).head(87)

Columns Datatypes
0            PROSPECTID     int64
1              Total_TL     int64
2         Tot_Closed_TL     int64
3         Tot_Active_TL     int64
4   Total_TL_opened_L6M     int64
..                  ...       ...
82              GL_Flag     int64
83       last_prod_enq2    object
84      first_prod_enq2    object
85         Credit_Score     int64
86        Approved_Flag    object

[87 rows x 2 columns]

In [13]:
# object colums need =['material_status','Education','gender','last_prod_loan','first_prod_loan','Approved_flag']
# int columns need =[All except object col]

In [14]:
final_df.select_dtypes('object').columns

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

In [15]:
final_df.select_dtypes('number').columns

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL',
       'time_since_recent_payment', 'time_since_first_deliquency',
       'time_since_recent_deliquency', 'num_times_delinquent',
       'max_delinquency_level', 'max_recent_level_of_deliq', 'num_deliq_6mts',
       'num_deliq_12mts', 'num_deliq_6_12mts', 'max_deliq_6mts',
       'max_deliq_12mts', 'num_times_30p_dpd', 'num_times_60p_dpd', 'num_std',
       'num_std_6mts', 'num_std_12mts', 'num_sub', 'num_sub_6mts',
       'num_sub_12mts', 'num_dbt', 'num_dbt_6mts', 'num_dbt_12mts', 'num_lss',
       'num_lss_6mts', 'n

# Conclussion
1. `Datatypes` should be in correct format.

# 6. Null Values
1. First we can make the copy of dataframe

In [16]:
copy_df=final_df.copy()

In [17]:
final_df.isnull().sum()

PROSPECTID             0
Total_TL               0
Tot_Closed_TL          0
Tot_Active_TL          0
Total_TL_opened_L6M    0
                      ..
GL_Flag                0
last_prod_enq2         0
first_prod_enq2        0
Credit_Score           0
Approved_Flag          0
Length: 87, dtype: int64

1. **We can't see missing values but in above we say that `-99999` shows the missing values**
2. **Check if the ratio of `-99999` is greater in the specific column remove the columns**

# Handling Missing Values

In [18]:
final_df[final_df['time_since_recent_payment']==-99999].shape[0]

4291

In [19]:
final_df.shape

(51336, 87)

1. **We have `51336` records we can set the treshold `10000` if any columns contain `-99999` more then 10000 we can drop the columns.** 
2. **Also check the ratio of containg `-99999`.**

In [20]:
col_to_drop=[]
for i in final_df.columns:
    temp=final_df[final_df[i]==-99999].shape[0]
    ratio= round(temp/len(final_df),3)*100
    if(temp>10000):
        print(f"{i} -> -99999 value is -> {temp} and ratio is {ratio}%")
        col_to_drop.append(i)

time_since_first_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
time_since_recent_deliquency -> -99999 value is -> 35949 and ratio is 70.0%
max_delinquency_level -> -99999 value is -> 35949 and ratio is 70.0%
max_deliq_6mts -> -99999 value is -> 12890 and ratio is 25.1%
max_deliq_12mts -> -99999 value is -> 10832 and ratio is 21.099999999999998%
CC_utilization -> -99999 value is -> 47636 and ratio is 92.80000000000001%
PL_utilization -> -99999 value is -> 44435 and ratio is 86.6%
max_unsec_exposure_inPct -> -99999 value is -> 23178 and ratio is 45.1%


# Conclussion
1. `time_since_first_deliquency` contain `70.0%` missing data.
2. `time_since_recent_deliquency` contain `70.0%` missing data.
3. `max_delinquency_level` contain `70.0%` missing data.
4. `max_deliq_6mts` contain `25.1%` missing data.
5. `max_deliq_12mts` contain  `21.09% missing data.
6. `CC_utilization` conrain `92.8%` missing data.
7. `PL_utilization` contain `86.6%` missing data.
8. `max_unsec_exposure_inPct` contain  `45.1%` missing data.
- **We will drop these columns**

### Drop Columns 

In [21]:
final_df.drop(columns=col_to_drop,inplace=True)

In [22]:
final_df.shape

(51336, 79)

In [23]:
final_df.sample(10)

PROSPECTID  Total_TL  Tot_Closed_TL  Tot_Active_TL  \
44443       44444         5              3              2   
31656       31657         2              0              2   
20853       20854         1              0              1   
34782       34783         2              2              0   
42862       42863         3              2              1   
10561       10562         2              1              1   
28956       28957         1              0              1   
32823       32824         7              0              7   
37331       37332         1              1              0   
11922       11923         1              1              0   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
44443                    1                  1            0.200   
31656                    1                  0            0.500   
20853                    0                  0            0.000   
34782                    0                  1            0.000   
42862                    0                  1            0.000   
10561                    0                  0            0.000   
28956                    0                  0            0.000   
32823                    4                  0            0.571   
37331                    0                  1            0.000   
11922                    0                  0            0.000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  \
44443              0.200          0.400          0.600  ...   
31656              0.000          1.000          0.000  ...   
20853              0.000          1.000          0.000  ...   
34782              0.500          0.000          1.000  ...   
42862              0.333          0.333          0.667  ...   
10561              0.000          0.500          0.500  ...   
28956              0.000          1.000          0.000  ...   
32823              0.000          1.000          0.000  ...   
37331              1.000          0.000          1.000  ...   
11922              0.000          0.000          1.000  ...   

       pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
44443                     0.0                     1.0                     0.0   
31656                     0.0                     0.0                     0.0   
20853                     0.0                     0.0                     0.0   
34782                     0.0                     0.0                     0.0   
42862                     0.0                     0.0                     0.0   
10561                     0.0                     0.0                     0.0   
28956                     0.0                     0.0                     0.0   
32823                     0.0                     0.0                     0.0   
37331                     0.0                     0.0                     0.0   
11922                     0.0                     0.0                     0.0   

       pct_CC_enq_L6m_of_ever  HL_Flag  GL_Flag  last_prod_enq2  \
44443                     1.0        0        0              CC   
31656                     0.0        0        0          others   
20853                     0.0        0        0          others   
34782                     0.0        0        0    ConsumerLoan   
42862                     0.0        0        0    ConsumerLoan   
10561                     0.0        0        0    ConsumerLoan   
28956                     0.0        0        0          others   
32823                     0.0        0        0    ConsumerLoan   
37331                     0.0        0        0              AL   
11922                     0.0        0        0              PL   

       first_prod_enq2  Credit_Score  Approved_Flag  
44443     ConsumerLoan           678             P2  
31656           others           675             P2  
20853           others           704             P1  
34782     ConsumerLoan           667             P3  
42862         

# Conclussion
- After removing columns which contain `-99999` values more then `10000` shape of the dataframe is `51336` rows and `79` columns.
- we can remove only those columns which contain `-99999` values more then `10000` but remember that `-99999` are also present in those columns which contain  `-99999` values less then `10000`.
# Point to remember
- we have 2 option for `handling the remaing missing values`.
    ## Remove Missing Values
    - first we can remove those value.
        - we can remove the missing values if we maintain the balance of data we can remove them.
    ## Fill Missing Values
    - Fill these values
        - Filling the missing values is a cretical thing but if we can loss lot of data after removing the missing values then we pick the filling missing values.

- **Try to remove rows that contain `-99999` value**

In [24]:
for i in final_df.columns:
    print(final_df[~(final_df[i]==-99999)].shape[0])
    final_df=final_df[~(final_df[i]==-99999)]

51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51336
51296
51296
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
47042
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42099
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064
42064


# Conclussion
- Check the shape after removing.
- Also check the ration before removing and after removing

In [25]:
print(f"Shape of the data after removing -9999 values is: {final_df.shape}")

Shape of the data after removing -9999 values is: (42064, 79)


In [26]:
len(final_df)/len(copy_df)*100

81.93860059217702

# Conclussion
- we see that after removing the rows that contain `-99999` values we can maintain `80%` data.
- Now we sholud pick remove missing value and we will remove them.

In [27]:
# final_df.to_csv('credit.csv',index=False)

# Duplicates

In [28]:
final_df.duplicated().sum()

0

# Conclussion
- There is no duplicate found in this dataset.

# Unique Values

In [29]:
for i in final_df.describe(include='object'):
    print("Unique valeu in: ",i)
    
    print(final_df[i].unique())
    
    print("Total Unique values: ",final_df[i].nunique())
    
    print("-"*50)
    print()

Unique valeu in:  MARITALSTATUS
['Married' 'Single']
Total Unique values:  2
--------------------------------------------------

Unique valeu in:  EDUCATION
['12TH' 'GRADUATE' 'SSC' 'POST-GRADUATE' 'UNDER GRADUATE' 'OTHERS'
 'PROFESSIONAL']
Total Unique values:  7
--------------------------------------------------

Unique valeu in:  GENDER
['M' 'F']
Total Unique values:  2
--------------------------------------------------

Unique valeu in:  last_prod_enq2
['PL' 'ConsumerLoan' 'AL' 'CC' 'others' 'HL']
Total Unique values:  6
--------------------------------------------------

Unique valeu in:  first_prod_enq2
['PL' 'ConsumerLoan' 'others' 'AL' 'HL' 'CC']
Total Unique values:  6
--------------------------------------------------

Unique valeu in:  Approved_Flag
['P2' 'P1' 'P3' 'P4']
Total Unique values:  4
--------------------------------------------------



# Conclussion
- There are `2` unique values in  `MARITALSTATUS` i-e  `['Married' 'Single']`
- There are `7` unique values in  `EDUCATION` i-e `['12TH' 'GRADUATE' 'SSC' 'POST-GRADUATE' 'UNDER GRADUATE' 'OTHERS'
 'PROFESSIONAL']`
- There are `2` unique values in  `GENDER` i-e `['M' 'F']`
- There are `6` unique values in  `last_prod_enq2` i-e `['PL' 'ConsumerLoan' 'AL' 'CC' 'others' 'HL']`
- There are `6` unique values in  `first_prod_enq2` i-e `['PL' 'ConsumerLoan' 'others' 'AL' 'HL' 'CC']`
- There are `4` unique values in  `Approved_Flag` i-e `['P2' 'P1' 'P3' 'P4']`

# Statistical Summary

In [30]:
final_df.describe()

PROSPECTID      Total_TL  Tot_Closed_TL  Tot_Active_TL  \
count  42064.000000  42064.000000   42064.000000   42064.000000   
mean   25649.827477      5.262980       2.967383       2.295597   
std    14844.173396      7.463383       6.141098       2.404086   
min        1.000000      1.000000       0.000000       0.000000   
25%    12776.750000      1.000000       0.000000       1.000000   
50%    25706.500000      3.000000       1.000000       2.000000   
75%    38518.250000      6.000000       3.000000       3.000000   
max    51336.000000    235.000000     216.000000      47.000000   

       Total_TL_opened_L6M  Tot_TL_closed_L6M  pct_tl_open_L6M  \
count         42064.000000        42064.00000     42064.000000   
mean              0.812643            0.48992         0.179032   
std               1.383559            1.05892         0.278043   
min               0.000000            0.00000         0.000000   
25%               0.000000            0.00000         0.000000   
50%               0.000000            0.00000         0.000000   
75%               1.000000            1.00000         0.333000   
max              27.000000           19.00000         1.000000   

       pct_tl_closed_L6M  pct_active_tl  pct_closed_tl  ...  \
count       42064.000000   42064.000000   42064.000000  ...   
mean            0.097783       0.577452       0.422548  ...   
std             0.210957       0.366110       0.366110  ...   
min             0.000000       0.000000       0.000000  ...   
25%             0.000000       0.286000       0.000000  ...   
50%             0.000000       0.545000       0.455000  ...   
75%             0.100000       1.000000       0.714000  ...   
max             1.000000       1.000000       1.000000  ...   

       pct_currentBal_all_TL       CC_Flag       PL_Flag  \
count           42064.000000  42064.000000  42064.000000   
mean                0.883693      0.102962      0.193063   
std                40.622275      0.303913      0.394707   
min                 0.000000      0.000000      0.000000   
25%                 0.152000      0.000000      0.000000   
50%                 0.600000      0.000000      0.000000   
75%                 0.860000      0.000000      0.000000   
max              6327.500000      1.000000      1.000000   

       pct_PL_enq_L6m_of_L12m  pct_CC_enq_L6m_of_L12m  pct_PL_enq_L6m_of_ever  \
count            42064.000000            42064.000000            42064.000000   
mean                 0.219169                0.074833                0.195497   
std                  0.395100                0.250658                0.367414   
min                  0.000000                0.000000                0.000000   
25%                  0.000000                0.000000                0.000000   
50%                  0.000000                0.000000                0.000000   
75%                  0.000000                0.000000                0.000000   
max                  1.000000                1.000000                1.000000   

       pct_CC_enq_L6m_of_ever       HL_Flag       GL_Flag  Credit_Score  
count            42064.000000  42064.000000  42064.000000  42064.000000  
mean                 0.064186      0.252235      0.056580    679.326336  
std                  0.225989      0.434300      0.231042     21.133619  
min                  0.000000      0.000000      0.000000    469.000000  
25%                  0.000000      0.000000      0.000000    668.000000  
50%                  0.000000      0.000000      0.000000    679.000000  
75%                  0.000000      1.000000      0.000000    690.000000  
max                  1.000000      1.000000      1.000000    809.000000  

[8 rows x 73 columns]

# Conclussion
- **We can't see clear information first we can do `Feature Engnering` and then we can see clear information.**

# Feature Engineering

- In Feature Engineering, we can reduce the dimensionality of data by applying some statistical tests.

## Hyphothesis Testing
- Hyphothesis test are the statistical test used to draw conclussion about poopulation based on sample data.
- Some statistical test are that are use in this dataset.

### Variance Influence Threshold
- Suitable for numerical input and numerical output.

### Chi-Square Test
- Explain more about the Chi-Square test here.

### ANOVA
- These tests are applied on categorical versus numerical columns. Note that categorical variables should have more than three categories.


# Saperate numerical and categorical col

In [31]:
cat_col=final_df.select_dtypes('object').columns
num_col=final_df.select_dtypes('number').columns

In [32]:
cat_col

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

In [33]:
num_col

Index(['PROSPECTID', 'Total_TL', 'Tot_Closed_TL', 'Tot_Active_TL',
       'Total_TL_opened_L6M', 'Tot_TL_closed_L6M', 'pct_tl_open_L6M',
       'pct_tl_closed_L6M', 'pct_active_tl', 'pct_closed_tl',
       'Total_TL_opened_L12M', 'Tot_TL_closed_L12M', 'pct_tl_open_L12M',
       'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'Auto_TL', 'CC_TL',
       'Consumer_TL', 'Gold_TL', 'Home_TL', 'PL_TL', 'Secured_TL',
       'Unsecured_TL', 'Other_TL', 'Age_Oldest_TL', 'Age_Newest_TL',
       'time_since_recent_payment', 'num_times_delinquent',
       'max_recent_level_of_deliq', 'num_deliq_6mts', 'num_deliq_12mts',
       'num_deliq_6_12mts', 'num_times_30p_dpd', 'num_times_60p_dpd',
       'num_std', 'num_std_6mts', 'num_std_12mts', 'num_sub', 'num_sub_6mts',
       'num_sub_12mts', 'num_dbt', 'num_dbt_6mts', 'num_dbt_12mts', 'num_lss',
       'num_lss_6mts', 'num_lss_12mts', 'recent_level_of_deliq', 'tot_enq',
       'CC_enq', 'CC_enq_L6m', 'CC_enq_L12m', 'PL_enq', 'PL_enq_L6m',
       'PL_enq_L12

# Variance Inflence Factore (VIF)
- Now we can apply the `VIF` on numerical columns to check the relationship of each column to every other col.
- If the 2 or more columns are highly corelated so we can take only one column so that i will reduce the `multicolarnarty`.
- If we can reduce multicolarty the column will automaticlally reduce.

# Apply VIF
- we can set the threshold ratio if the variance is greater then treshold we can drop the columns.
- if the variance is less then treshold we can keep the columns.

In [34]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

In [35]:
col_to_kept=[]
vif_data=final_df[num_col]
column_index=0
for i in range(num_col.shape[0]):
    # variance_inflation_factor(dataframe,index)
    vif_value=variance_inflation_factor(vif_data,column_index)
    print(i,"------->",vif_value)
    
    if(vif_value<6):
        col_to_kept.append(num_col[i])
        column_index+=1
    else:
        vif_data=vif_data.drop(columns=num_col[i])

0 -------> 1.0014714453751743
1 -------> inf
2 -------> inf
3 -------> 11.320645246142803
4 -------> 8.364064272336133
5 -------> 6.520792569522158
6 -------> 5.1498060557107745
7 -------> 2.6112073771674944
8 -------> inf
9 -------> 1789.5578922792413
10 -------> 8.601096683974646
11 -------> 3.832851362174614
12 -------> 6.099768198336671
13 -------> 5.581446751193094
14 -------> 1.985655138966978
15 -------> inf
16 -------> 4.809807520859604
17 -------> 23.27288184101288
18 -------> 30.596519319624697
19 -------> 4.3848122091115815
20 -------> 3.06467356246316
21 -------> 2.8986527303140033
22 -------> 4.377945210179215
23 -------> 2.207856265954658
24 -------> 4.9200182258988345
25 -------> 5.215354852494453
26 -------> 3.3862072703025055
27 -------> 7.841101396160445
28 -------> 5.255263829126264
29 -------> inf
30 -------> 7.380644570758274
31 -------> 1.421011976182927
32 -------> 8.083354471350239
33 -------> 1.6241477619270308
34 -------> 7.257891375429919
35 -------> 15.59625

In [36]:
len(col_to_kept)

40

# Conclussion
- After applying `variance inflence factor` we can reduce some column now the remaing `numerical column is 40`
- we see that after applying `VIF` remaing numerical col is 40 now check the relationship of remaing numerical col to target column.

# ANOVA
- It is a hypothesis test in this test we can check the relationship b/w numerical col to target col.
- Target  col should be categorical and must contain 3 or greater then 3 categories.
- In our case our target col is `Approved_Flag` which contain `4 classes`.
- We can set the threshold `0.05` if the p_value is greater then `0.05` we can drop these columns.
- `0.05` threshold are generaly industry standard.

In [37]:
from scipy.stats import f_oneway

In [38]:
cat_col[-1]

'Approved_Flag'

In [39]:
col_to_remaing=[]
for i in col_to_kept:
    current_col=final_df[i]
    targte_col=final_df[cat_col[-1]]
    
    # Saperat each group for preforming ANOVA 
    group1=[value for value,group in zip(current_col,targte_col) if group=='P1']
    group2=[value for value,group in zip(current_col,targte_col) if group=='P2']
    group3=[value for value,group in zip(current_col,targte_col) if group=='P3']
    group4=[value for value,group in zip(current_col,targte_col) if group=='P4']
    
    # Perform ANOVA
    f_stats,p_value=f_oneway(group1,group2,group3,group4)
    print(i,"------>",p_value)
    
    if(p_value<0.05):
        col_to_remaing.append(i)

PROSPECTID ------> 0.04216738785382283
pct_tl_open_L6M ------> 0.0
pct_tl_closed_L6M ------> 2.676086811190998e-52
Tot_TL_closed_L12M ------> 1.762685578230421e-99
pct_tl_closed_L12M ------> 4.1384568281483056e-49
Tot_Missed_Pmnt ------> 9.572404615942641e-210
CC_TL ------> 5.710261952999025e-57
Home_TL ------> 0.0
PL_TL ------> 7.991892389241125e-235
Secured_TL ------> 0.0
Unsecured_TL ------> 2.102384431982427e-118
Other_TL ------> 0.0
Age_Oldest_TL ------> 0.0
Age_Newest_TL ------> 1.0465965445912546e-88
time_since_recent_payment ------> 8.943912807416738e-81
max_recent_level_of_deliq ------> 6.699932727692095e-81
num_deliq_6_12mts ------> 2.0772229883600566e-59
num_times_60p_dpd ------> 3.3833576401764997e-59
num_std_12mts ------> 0.0
num_sub ------> 1.6985545450345584e-22
num_sub_6mts ------> 0.016579119048249723
num_sub_12mts ------> 0.001299484021242129
num_dbt ------> 3.135918708089289e-11
num_dbt_12mts ------> 0.005446087209172642
num_lss ------> 0.009340555562870471
num_lss_1

In [40]:
len(col_to_remaing)

38

# Conclussion
- After `ANOVA` we can get only 38 columns.

# Chie-Square
- These test are generally perform on categorical column.
- Note that target column category should be `>=3`.
- Now using `chie-square` test we can check the relationship b/w target col.

In [41]:
from scipy.stats import chi2_contingency,chisquare

In [42]:
drop_col=[]
for i in cat_col[:-1]:
    # Perform chie-square
    chie,p_value,_,_=chi2_contingency(pd.crosstab(final_df[i],final_df[cat_col[-1]]))
    
    print(i,"----->",p_value)
    if(p_value>0.05):
        drop_col.append(i)

MARITALSTATUS -----> 3.578180861038862e-233
EDUCATION -----> 2.6942265249737532e-30
GENDER -----> 1.907936100186563e-05
last_prod_enq2 -----> 0.0
first_prod_enq2 -----> 7.84997610555419e-287


In [43]:
drop_col

[]

# Conclussion
- See that all the columns `p_value` is less the `0.05` so we can accept all the columns.

In [44]:
col_to_remaing

['PROSPECTID',
 'pct_tl_open_L6M',
 'pct_tl_closed_L6M',
 'Tot_TL_closed_L12M',
 'pct_tl_closed_L12M',
 'Tot_Missed_Pmnt',
 'CC_TL',
 'Home_TL',
 'PL_TL',
 'Secured_TL',
 'Unsecured_TL',
 'Other_TL',
 'Age_Oldest_TL',
 'Age_Newest_TL',
 'time_since_recent_payment',
 'max_recent_level_of_deliq',
 'num_deliq_6_12mts',
 'num_times_60p_dpd',
 'num_std_12mts',
 'num_sub',
 'num_sub_6mts',
 'num_sub_12mts',
 'num_dbt',
 'num_dbt_12mts',
 'num_lss',
 'recent_level_of_deliq',
 'CC_enq_L12m',
 'PL_enq_L12m',
 'time_since_recent_enq',
 'enq_L3m',
 'NETMONTHLYINCOME',
 'Time_With_Curr_Empr',
 'CC_Flag',
 'PL_Flag',
 'pct_PL_enq_L6m_of_ever',
 'pct_CC_enq_L6m_of_ever',
 'HL_Flag',
 'GL_Flag']

In [45]:
cat_col

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2', 'Approved_Flag'],
      dtype='object')

# Conclussion
- In `col to kept remaing` contain only numerical col.
- `cat-col` conatin  only categorical col.
- Both are saperate now first combine them into a single list.

In [46]:
col_to_remaing.extend(cat_col)

In [47]:
len(col_to_remaing)

44

- **we see that out of `88` columns we can get `44` column that met the conditions.**

In [48]:
final_df=final_df[col_to_remaing]
final_df.head()

PROSPECTID  pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
0           1            0.000                0.0                   0   
1           2            0.000                0.0                   0   
2           3            0.125                0.0                   0   
4           5            0.000                0.0                   0   
5           6            0.000                0.0                   1   

   pct_tl_closed_L12M  Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  \
0               0.000                0      0        0      4           1   
1               0.000                0      0        0      0           0   
2               0.000                1      0        0      0           2   
4               0.000                0      0        0      0           3   
5               0.167                0      0        0      0           6   

   ...  pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  HL_Flag  GL_Flag  \
0  ...                   0.000                     0.0        1        0   
1  ...                   0.000                     0.0        0        0   
2  ...                   0.000                     0.0        1        0   
4  ...                   0.000                     0.0        0        0   
5  ...                   0.429                     0.0        1        0   

   MARITALSTATUS      EDUCATION  GENDER  last_prod_enq2  first_prod_enq2  \
0        Married           12TH       M              PL               PL   
1         Single       GRADUATE       F    ConsumerLoan     ConsumerLoan   
2        Married            SSC       M    ConsumerLoan           others   
4        Married  POST-GRADUATE       M              AL               AL   
5        Married           12TH       M    ConsumerLoan               PL   

   Approved_Flag  
0             P2  
1             P2  
2             P2  
4             P1  
5             P3  

[5 rows x 44 columns]

In [49]:
final_df.shape

(42064, 44)

# Conclussion
- After Peprocessing we will get `42064` out of `51000` and `44` columns out of `88`

# Model Building
- First we can saperate the input and output columns.
- Second we can saperate numerical and categorical columns.
- Third we can build a pipeline for some transformation.

# Saperate Input and output col

In [50]:
feature=final_df.drop(columns=['Approved_Flag'])
label=final_df['Approved_Flag']

In [51]:
feature.head(2)

PROSPECTID  pct_tl_open_L6M  pct_tl_closed_L6M  Tot_TL_closed_L12M  \
0           1              0.0                0.0                   0   
1           2              0.0                0.0                   0   

   pct_tl_closed_L12M  Tot_Missed_Pmnt  CC_TL  Home_TL  PL_TL  Secured_TL  \
0                 0.0                0      0        0      4           1   
1                 0.0                0      0        0      0           0   

   ...  PL_Flag  pct_PL_enq_L6m_of_ever  pct_CC_enq_L6m_of_ever  HL_Flag  \
0  ...        1                     0.0                     0.0        1   
1  ...        0                     0.0                     0.0        0   

   GL_Flag  MARITALSTATUS  EDUCATION  GENDER  last_prod_enq2  first_prod_enq2  
0        0        Married       12TH       M              PL               PL  
1        0         Single   GRADUATE       F    ConsumerLoan     ConsumerLoan  

[2 rows x 43 columns]

In [52]:
label.head()

0    P2
1    P2
2    P2
4    P1
5    P3
Name: Approved_Flag, dtype: object

In [53]:
label.value_counts().sort_index()

Approved_Flag
P1     4908
P2    25452
P3     6440
P4     5264
Name: count, dtype: int64

# Encode target Colummn

In [54]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder,StandardScaler,OrdinalEncoder

In [55]:
lb=LabelEncoder()
label=lb.fit_transform(label)

In [56]:
label

array([1, 1, 1, ..., 2, 1, 1])

# Saperate `Numerical` and `Categorical` column

In [75]:
num_col=feature.select_dtypes('number').columns
cat_col=feature.select_dtypes('object').columns

In [76]:
num_col

Index(['PROSPECTID', 'pct_tl_open_L6M', 'pct_tl_closed_L6M',
       'Tot_TL_closed_L12M', 'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL',
       'Home_TL', 'PL_TL', 'Secured_TL', 'Unsecured_TL', 'Other_TL',
       'Age_Oldest_TL', 'Age_Newest_TL', 'time_since_recent_payment',
       'max_recent_level_of_deliq', 'num_deliq_6_12mts', 'num_times_60p_dpd',
       'num_std_12mts', 'num_sub', 'num_sub_6mts', 'num_sub_12mts', 'num_dbt',
       'num_dbt_12mts', 'num_lss', 'recent_level_of_deliq', 'CC_enq_L12m',
       'PL_enq_L12m', 'time_since_recent_enq', 'enq_L3m', 'NETMONTHLYINCOME',
       'Time_With_Curr_Empr', 'CC_Flag', 'PL_Flag', 'pct_PL_enq_L6m_of_ever',
       'pct_CC_enq_L6m_of_ever', 'HL_Flag', 'GL_Flag'],
      dtype='object')

In [77]:
cat_col

Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2'],
      dtype='object')

# Train Test Split

In [93]:
from sklearn.model_selection import train_test_split,cross_val_score

In [79]:
x_train,x_test,y_train,y_test=train_test_split(feature,label,test_size=0.2,random_state=43)

In [80]:
x_train.shape

(33651, 43)

In [81]:
x_test.shape

(8413, 43)

# Building Pipeline
- we can build a pipeline for numercial column.
    - In numerical pipeline we can scale the feature.
- we can build a pipeline for categorical column.
    - In categorical pipeline we can encode the categorical column.

In [82]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# Numerical Pipeline

In [83]:
num_pipe=Pipeline(steps=[
    ("impute",SimpleImputer(strategy='median')),
    ('scale',StandardScaler())
])
num_pipe

Pipeline(steps=[('impute', SimpleImputer(strategy='median')),
                ('scale', StandardScaler())])

# Categorical Pipeline

In [84]:
cat_pipe=Pipeline(steps=[
    ("impute",SimpleImputer(strategy='most_frequent')),
    ('Encode',OneHotEncoder(drop='first',sparse=False,handle_unknown='ignore'))
])
cat_pipe

Pipeline(steps=[('impute', SimpleImputer(strategy='most_frequent')),
                ('Encode',
                 OneHotEncoder(drop='first', handle_unknown='ignore',
                               sparse=False))])

# Colum Transformer
- In column transformer we can combine al the steps.

In [85]:
from sklearn.compose import ColumnTransformer

In [86]:
transformer=ColumnTransformer(transformers=[
    ('num_transformer',num_pipe,num_col),
    ("Encode",cat_pipe,cat_col)
],remainder='passthrough')
transformer

ColumnTransformer(remainder='passthrough',
                  transformers=[('num_transformer',
                                 Pipeline(steps=[('impute',
                                                  SimpleImputer(strategy='median')),
                                                 ('scale', StandardScaler())]),
                                 Index(['PROSPECTID', 'pct_tl_open_L6M', 'pct_tl_closed_L6M',
       'Tot_TL_closed_L12M', 'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL',
       'Home_TL', 'PL_TL', 'Secured_TL', 'Unsecured_TL', 'Other_TL',
       'Ag...
       'Time_With_Curr_Empr', 'CC_Flag', 'PL_Flag', 'pct_PL_enq_L6m_of_ever',
       'pct_CC_enq_L6m_of_ever', 'HL_Flag', 'GL_Flag'],
      dtype='object')),
                                ('Encode',
                                 Pipeline(steps=[('impute',
                                                  SimpleImputer(strategy='most_frequent')),
                                                 ('Encode',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore',
                                                                sparse=False))]),
                                 Index(['MARITALSTATUS', 'EDUCATION', 'GENDER', 'last_prod_enq2',
       'first_prod_enq2'],
      dtype='object'))])

# Build Final Pipeline for model

# Fit the model
- In this we can train different models.
- Those model will give better result we can tune them.

In [87]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
import xgboost as xgb
from xgboost import XGBClassifier

# Make the models dict

In [88]:
model_dic={
    "LogisticRegression":LogisticRegression(),
    "RandomForest":RandomForestClassifier(),
    "DecessionTree":DecisionTreeClassifier(),
    "xgboost":XGBClassifier()
}

# Fit models
- Fit models and also calculate the score

In [89]:
from sklearn.metrics import accuracy_score,confusion_matrix

In [96]:
results={
    "model_name":[],
    "score":[],
    'train score':[],
    "test score":[],
    'matrix':[]
}
for model_name,model in model_dic.items():
    final=Pipeline(steps=[
    ("process",transformer),
    ("model",model)
    ])
    
    # Fit the model pipe
    final.fit(x_train,y_train)
    
    # Prediction
    pre=final.predict(x_test)
    
    # Calculate score
    score=accuracy_score(y_test,pre)
    matrix=confusion_matrix(y_test,pre)
    
    #Cross val score
    train_score=cross_val_score(final,x_train,y_train,cv=5,scoring='accuracy')
    test_score=cross_val_score(final,x_test,y_test,cv=5,scoring='accuracy')
    
    
    results['model_name'].append(model_name)
    results['score'].append(score)
    results['train score'].append(train_score)
    results['test score'].append(test_score)
    
    results['matrix'].append(matrix)

# See the results

In [97]:
pd.DataFrame(results)

model_name     score  \
0  LogisticRegression  0.754784   
1        RandomForest  0.772376   
2       DecessionTree  0.703792   
3             xgboost  0.785094   

                                         train score  \
0  [0.7520427871044422, 0.7533432392273403, 0.749...   
1  [0.7671965532610311, 0.7647845468053491, 0.763...   
2  [0.6966275441984846, 0.6968796433878157, 0.695...   
3  [0.7768533650274848, 0.7732540861812779, 0.770...   

                                          test score  \
0  [0.750445632798574, 0.7474747474747475, 0.7403...   
1  [0.7617349970291146, 0.7635175282234106, 0.761...   
2  [0.7035056446821153, 0.696969696969697, 0.6815...   
3  [0.7575757575757576, 0.7706476530005941, 0.770...   

                                              matrix  
0  [[685, 280, 2, 3], [149, 4841, 88, 74], [27, 8...  
1  [[704, 266, 0, 0], [133, 4802, 170, 47], [30, ...  
2  [[678, 245, 46, 1], [264, 4126, 638, 124], [31...  
3  [[765, 205, 0, 0], [148, 4697, 260, 47], [31, ...

# Confussion Matrix

## Logistic Regression 

In [99]:
results['matrix'][0]

array([[ 685,  280,    2,    3],
       [ 149, 4841,   88,   74],
       [  27,  865,  158,  217],
       [   0,  265,   93,  666]], dtype=int64)

# Random Forest

In [103]:
results['matrix'][1]

array([[ 704,  266,    0,    0],
       [ 133, 4802,  170,   47],
       [  30,  754,  269,  214],
       [   2,  154,  145,  723]], dtype=int64)

# Desession Tree

In [104]:
results['matrix'][2]

array([[ 678,  245,   46,    1],
       [ 264, 4126,  638,  124],
       [  31,  527,  447,  262],
       [   2,  117,  235,  670]], dtype=int64)

# XgBoost

In [105]:
results['matrix'][3]

array([[ 765,  205,    0,    0],
       [ 148, 4697,  260,   47],
       [  31,  633,  392,  211],
       [   0,   79,  194,  751]], dtype=int64)

# Conclussion
- We see that `Xgboost` give better result with `accuracy score` of `78` now we can fine tune them

In [128]:
pd.Series(y_test).value_counts().sort_index()

0     970
1    5152
2    1267
3    1024
Name: count, dtype: int64

# Hyper Parameter Tunning

In [129]:
from sklearn.model_selection import GridSearchCV

In [145]:
param_grid = {
    'model__model__colsample_bytree': [0.1, 0.3, 0.5, 0.7, 0.9],
    'model__model__learning_rate': [0.001, 0.01, 0.1, 1],
    'model__model__max_depth': [3, 5, 8, 10],
    'model__model__n_estimators': [10, 50, 100]
}

In [146]:
xgb_boost=XGBClassifier()
pipeline=Pipeline(steps=[
    ("process",transformer),
    ('model',xgb_boost)
])

In [149]:
grid_search=GridSearchCV(pipeline,param_grid=param_grid,scoring='accuracy',n_jobs=-1,cv=5)
grid_search.fit(x_train,y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('process',
                                        ColumnTransformer(remainder='passthrough',
                                                          transformers=[('num_transformer',
                                                                         Pipeline(steps=[('impute',
                                                                                          SimpleImputer(strategy='median')),
                                                                                         ('scale',
                                                                                          StandardScaler())]),
                                                                         Index(['PROSPECTID', 'pct_tl_open_L6M', 'pct_tl_closed_L6M',
       'Tot_TL_closed_L12M', 'pct_tl_closed_L12M', 'Tot_Missed_Pmnt', 'CC_TL',
       'Home...
                                                      missing=nan,
                                                      monotone_constraints=None,
                                                      multi_strategy=None,
                                                      n_estimators=None,
                                                      n_jobs=None,
                                                      num_parallel_tree=None,
                                                      random_state=None, ...))]),
             n_jobs=-1,
             param_grid={'model__model__colsample_bytree': [0.1, 0.3, 0.5, 0.7,
                                                            0.9],
                         'model__model__learning_rate': [0.001, 0.01, 0.1, 1],
                         'model__model__max_depth': [3, 5, 8, 10],
                         'model__model__n_estimators': [10, 50, 100]},
             scoring='accuracy')

In [150]:
grid_search.best_params_

{'model__model__colsample_bytree': 0.1,
 'model__model__learning_rate': 0.001,
 'model__model__max_depth': 3,
 'model__model__n_estimators': 10}

In [152]:
pre=grid_search.predict(x_test)
pre

array([1, 2, 1, ..., 3, 1, 1], dtype=int64)

# Check accuracy again

In [153]:
accuracy_score(y_test,pre)

0.7850944966123856

# confussion Matrix

In [159]:
confusion_matrix(y_test,pre)

array([[ 765,  205,    0,    0],
       [ 148, 4697,  260,   47],
       [  31,  633,  392,  211],
       [   0,   79,  194,  751]], dtype=int64)

# Corss Validation

In [161]:
train_score=cross_val_score(pipeline,x_train,y_train,cv=5,scoring='accuracy').mean()
test_score=cross_val_score(pipeline,x_test,y_test,cv=5,scoring='accuracy').mean()

In [162]:
print("Train score is: ",train_score)
print("Test Score is: ",test_score)

Train score is:  0.7754301083695385
Test Score is:  0.7669086472192019
